File: examples/durability/retry-failure/java/RetryFailureAgent.java
package ai.axme.examples.durability;

import ai.axme.examples.SseHelper;
import dev.axme.sdk.*;
import java.security.MessageDigest;
import java.util.*;

/** Webhook Receiver Agent — durability/retry-failure (Java). */
public class RetryFailureAgent {
    static final String BASE = SseHelper.env("AXME_BASE_URL", "https://api.cloud.axme.ai");
    static final String KEY = System.getenv("AXME_API_KEY");
    static final String ADDR = SseHelper.env("AXME_AGENT_ADDRESS", "webhook-receiver-agent");

    public static void main(String[] args) throws Exception {
        if (KEY == null || KEY.isEmpty()) { System.err.println("AXME_API_KEY is required"); System.exit(1); }
        AxmeClient client = new AxmeClient(new AxmeClientConfig(BASE, KEY));
        System.out.printf("webhook-receiver starting  address=%s%n", ADDR);
        int since = 0;
        while (true) {
            try {
                for (Map<String, Object> d : SseHelper.pollAgentStream(BASE, KEY, ADDR, since, 15)) {
                    String id = SseHelper.str(d, "intent_id"); Number seq = (Number) d.get("seq");
                    if (seq != null) since = Math.max(since, seq.intValue()); if (id.isEmpty()) continue;
                    Map<String, Object> intent = SseHelper.unwrapIntent(client.getIntent(id, RequestOptions.none()));
                    if (!SseHelper.isActionable(intent)) continue;
                    Map<String, Object> p = SseHelper.effectivePayload(intent);

                    String et = SseHelper.str(p, "event_type", "unknown"), src = SseHelper.str(p, "source");
                    String tid = SseHelper.str(p, "test_id");
                    MessageDigest md = MessageDigest.getInstance("SHA-256");
                    String fp = bytesToHex(md.digest((et + ":" + src + ":" + tid).getBytes())).substring(0, 16);

                    client.resumeIntent(id, Map.of("action", "complete", "event_type", et, "source", src,
                            "fingerprint", fp, "processed", true, "note", "webhook processed"), SseHelper.withOwner(ADDR));
                    System.out.printf("resumed %s fingerprint=%s%n", id, fp);
                }
            } catch (Exception e) { Thread.sleep(2000); }
        }
    }

    static String bytesToHex(byte[] b) {
        StringBuilder sb = new StringBuilder();
        for (byte v : b) sb.append(String.format("%02x", v));
        return sb.toString();
    }
}

File: examples/dotnet/Durability/RetryFailureAgent.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;
using Axme.Sdk;
using AxmeExamples;

var baseUrl = SseHelper.Env("AXME_BASE_URL", "https://api.cloud.axme.ai");
var apiKey = SseHelper.Env("AXME_API_KEY");
var addr = SseHelper.Env("AXME_AGENT_ADDRESS", "webhook-receiver-agent");
if (string.IsNullOrEmpty(apiKey)) { Console.Error.WriteLine("AXME_API_KEY is required"); return; }

var client = new AxmeClient(new AxmeClientConfig { ApiKey = apiKey, BaseUrl = baseUrl });
Console.WriteLine($"webhook-receiver starting  address={addr}");
var since = 0;

while (true)
{
    try
    {
        foreach (var d in await SseHelper.PollAgentStreamAsync(baseUrl, apiKey, addr, since))
        {
            var id = SseHelper.Str(d, "intent_id");
            if (d["seq"]?.GetValue<int>() is int seq) since = Math.Max(since, seq);
            if (string.IsNullOrEmpty(id)) continue;
            var intent = SseHelper.UnwrapIntent(await client.GetIntentAsync(id));
            if (!SseHelper.IsActionable(intent)) continue;
            var p = SseHelper.EffectivePayload(intent);

            var et = SseHelper.Str(p, "event_type", "unknown"); var src = SseHelper.Str(p, "source");
            var tid = SseHelper.Str(p, "test_id");
            var fp = Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes($"{et}:{src}:{tid}"))).ToLower()[..16];

            var res = new JsonObject { ["action"] = "complete", ["event_type"] = et, ["source"] = src,
                ["fingerprint"] = fp, ["processed"] = true, ["note"] = $"webhook '{et}' processed" };
            await client.ResumeIntentAsync(id, res, new RequestOptions { OwnerAgent = addr });
            Console.WriteLine($"resumed {id} fingerprint={fp}");
        }
    }
    catch { await Task.Delay(2000); }
}
